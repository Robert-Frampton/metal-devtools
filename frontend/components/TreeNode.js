import Component, {Config} from 'metal-jsx';
import {bindAll, isEqual} from 'lodash';

import NodeName, {OPENING, NORMAL_CLOSING, SELF_CLOSING} from './NodeName';

class TreeNode extends Component {
	created() {
		bindAll(
			this,
			'focusNode',
			'toggleExpanded'
		);

		this._firstRender = true;
	}

	focusNode() {
		const {componentNode, onNodeClick} = this.props;

		onNodeClick(componentNode.id);
	}

	toggleExpanded(event) {
		event.stopPropagation();

		const {componentNode, onNodeClick} = this.props;

		const newVal = !this.state.expanded_;

		this.state.expanded_ = newVal;

		onNodeClick(componentNode.id);
	}

	toggleHighlight(value) {
		return () => {
			this.state.highlight_ = value;
		};
	}

	syncComponentNode(newVal = {}, oldVal = {}) {
		const newData = newVal.data || {};
		const oldData = oldVal.data || {};

		if (!this._firstRender && !isEqual(newData, oldData)) {
			const {element} = this.refs.nodeName;

			element.classList.add('flash');

			setTimeout(
				() => {
					element.classList.remove('flash');
				},
				100
			);
		}
		else if (this._firstRender) {
			this._firstRender = false;
		}
	}

	rendered() {
		if (this.props.componentNode.containsInspected) {
			this.state.expanded_ = true;

			delete this.props.componentNode.containsInspected;
		}
	}

	render() {
		const {componentNode, depth, onNodeClick, selectedId} = this.props;

		const {childComponents, id, name} = componentNode;

		const {expanded_, highlight_} = this.state;

		const hasChildren = childComponents && childComponents.length > 0;

		const selected = id === selectedId ? 'selected' : '';

		const highlight = highlight_ ? 'highlight' : '';

		const style = `padding-left: ${depth * 24 + 12}px`;

		return(
			<div class="tree-container">
				<div
					class={`node-wrapper ${selected} ${highlight} ${hasChildren ? 'expandable' : ''}`}
					onClick={expanded_ ? this.focusNode : this.toggleExpanded}
					onMouseEnter={this.toggleHighlight(true)}
					onMouseLeave={this.toggleHighlight(false)}
					style={style}
				>
					{hasChildren &&
						<div class={expanded_ ? 'arrow down' : 'arrow right'} onClick={this.toggleExpanded} />
					}

					<NodeName ref="nodeName" name={name} type={hasChildren ? OPENING : SELF_CLOSING} />
				</div>

				{hasChildren && expanded_ &&
					childComponents.map(
						(child, i) => (
							<TreeNode
								componentNode={child}
								depth={this.props.depth + 1}
								key={`${name}-${i}`}
								onNodeClick={onNodeClick}
								selectedId={selectedId}
							/>
						)
					)
				}

				{hasChildren && expanded_ &&
					<div
						class={`node-wrapper ${selected} ${highlight}`}
						onClick={this.focusNode}
						onMouseEnter={this.toggleHighlight(true)}
						onMouseLeave={this.toggleHighlight(false)}
						style={style}
					>
						<NodeName name={name} type={NORMAL_CLOSING} />
					</div>
				}
			</div>
		);
	}
}

TreeNode.PROPS = {
	componentNode: Config.value({}),
	depth: Config.number().value(0),
	onNodeClick: Config.func(),
	selectedId: Config.string()
};

TreeNode.STATE = {
	expanded_: Config.value(false),
	highlight_: Config.value(false)
};

export default TreeNode;
