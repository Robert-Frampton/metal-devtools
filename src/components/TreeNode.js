import Component, {Config} from 'metal-jsx';

const collapsedArrow = '+';
const expandedArrow = '-';

class TreeNode extends Component {
	created() {
		this.toggleExpanded = this.toggleExpanded.bind(this);
	}

	toggleExpanded() {
		const newVal = !this.state.expanded_;

		this.state.expanded_ = newVal;

		this.props.componentNode._expanded = newVal;
	}

	render() {
		const {
			childComponents,
			_expanded,
			name
		} = this.props.componentNode;

		const expanded = _expanded || this.state.expanded_;

		const arrow = (expanded) ? expandedArrow : collapsedArrow;

		const hasChildren = childComponents && childComponents.length > 0;

		const empty = hasChildren ? '' : 'empty';

		return(
			<div class="tree">
				<div class="tree-node" onClick={this.toggleExpanded}>
					{hasChildren &&
						<div class="toggle">
							{arrow}
						</div>
					}

					<div class={`component-info ${empty}`}>
						<span>{'<'}</span>

						{name}

						<span>{'>'}</span>
					</div>
				</div>

				{hasChildren && expanded &&
					childComponents.map(
						(child, i) => <TreeNode componentNode={child} key={`${name}-${i}`} />
					)
				}
			</div>
		)
	}
}

TreeNode.PROPS = {
	componentNode: Config.value({}),
	expanded: Config.value(false)
};

TreeNode.STATE = {
	expanded_: Config.value(false)
};

export default TreeNode;
