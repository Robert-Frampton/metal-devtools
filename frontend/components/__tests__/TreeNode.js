jest.unmock('../TreeNode');
jest.unmock('../NodeName');

import TreeNode from '../TreeNode';

describe('TreeNode', () => {
	it('should render', () => {
		const component = new TreeNode();

		expect(component.element).toBeTruthy();
	});
});
