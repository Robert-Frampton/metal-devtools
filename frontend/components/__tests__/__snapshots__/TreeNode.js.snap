exports[`TreeNode should render 1`] = `
<div class="tree-container">
  <div class="node-wrapper selected  expandable" style="padding-left: 12px;">
    <div class="arrow right"></div>
    <span class="node-name-container">&lt;
      <span class="name"></span>&gt;</span>
  </div>
</div>
`;

exports[`TreeNode should show when expanded 1`] = `
<div class="tree-container">
  <div class="node-wrapper selected  expandable" style="padding-left: 12px;">
    <div class="arrow down"></div>
    <span class="node-name-container">&lt;
      <span class="name"></span>&gt;</span>
  </div>
  <div class="tree-container">
    <div class="node-wrapper selected  " style="padding-left: 36px;">
      <span class="node-name-container">&lt;
        <span class="name"></span> /&gt;</span>
    </div>
  </div>
  <div class="node-wrapper " style="padding-left: 12px;">
    <span class="node-name-container">&lt;/
      <span class="name"></span>&gt;</span>
  </div>
</div>
`;

exports[`TreeNode should show when highlighted 1`] = `
<div class="tree-container">
  <div class="node-wrapper selected  expandable" style="padding-left: 12px;">
    <div class="arrow right"></div>
    <span class="node-name-container">&lt;
      <span class="name"></span>&gt;</span>
  </div>
</div>
`;
